File: trunk/tests/ccsp-comparisons/erlang/agents.erl
-module(agents).
-export([main/0, main/1, agent/3, barrier_control/3, location/1, view/3]).

-import(lists, [keytake/3, nth/2, reverse/1, unzip/1]).

-define(N_CYCLES, 1024).
-define(WORLD_SIZE, 5).
-define(LOC_AGENTS, 12).
-define(LOC_SIZE, 4096).
-define(LOC_AREA, ?LOC_SIZE * ?LOC_SIZE).
-define(LOC_MAX, (?LOC_SIZE div 2) - 1).
-define(LOC_MIN, -(?LOC_SIZE div 2)).
-define(OFFSETS,
  [{-1, -1}, {0, -1}, {1, -1}, {-1, 0}, {1, 0}, {-1, 1}, {0, 1}, {1, 1}, {0, 0}]
).

%% Vectors

vector_add({X1, Y1}, {X2, Y2}) -> {X1 + X2, Y1 + Y2}.
vector_sub({X1, Y1}, {X2, Y2}) -> {X1 - X2, Y1 - Y2}.


%% Barrier

release_waiting([]) ->
  nop;
release_waiting([P|PS]) ->
  P ! sync,
  release_waiting(PS).

barrier_control(Enrolled, Count, Waiting) ->
  receive
    {sync, P} -> 
      if
        Count > 1 ->
	  barrier_control(Enrolled, Count - 1, [P|Waiting]);
	true      ->
	  release_waiting([P|Waiting]),
	  barrier_control(Enrolled, Enrolled, [])
      end;
    {enroll, N} ->
      barrier_control(Enrolled + N, Count + N, Waiting);
    {resign, N} ->
      if
        (Count - N) == 0 ->
	  release_waiting(Waiting),
	  barrier_control(Enrolled - N, Enrolled - N, []);
	true ->
	  barrier_control(Enrolled - N, Count - N, Waiting)
      end;
    shutdown ->
      nop
  end.

init_barrier(Count) -> 
  register(barrier, spawn(agents, barrier_control, [Count, Count, []])).

sync_barrier() ->
  barrier ! {sync, self()},
  receive
    sync ->
      nop
  end.

shutdown_barrier() ->
  barrier ! shutdown.


%% Locations and Views

make_nv({X, Y}) -> {X * ?LOC_SIZE, Y * ?LOC_SIZE}.
get_neighbour(N, Neighbourhood) -> element(N + 1, Neighbourhood).

dir_value(N) ->
  if
    N < ?LOC_MIN -> -1;
    N > ?LOC_MAX -> 1;
    true -> 0
  end.

find_offset(V, N, []) ->
  {N, V};
find_offset(V, N, [O|OS]) ->
  if
    V == O -> {N, O};
    true   -> find_offset(V, (N + 1), OS)
  end.

calculate_direction({X, Y}) ->
  find_offset({dir_value(X), dir_value(Y)}, 0, ?OFFSETS).

in_bounds(Vector) ->
  {X, Y} = Vector,
  if
    X < ?LOC_MIN; X > ?LOC_MAX; Y < ?LOC_MIN; Y > ?LOC_MAX -> false;
    true -> true
  end.

adjust_agents(_, [], Result) ->
  Result;
adjust_agents(V, [A|AS], Result) ->
  {ID, Loc, Pos} = A,
  NewPos = vector_add(Pos, V),
  adjust_agents(V, AS, [{ID, Loc, NewPos}|Result]).

build_view([], [], Agents) ->
  Agents;
build_view([O|OS], [L|LS], Agents) -> 
  L ! {get_info, self()},
  receive
    {info, AgentList} ->
      NewAgents = adjust_agents(make_nv(O), AgentList, Agents),
      build_view(OS, LS, NewAgents)
  end.

view(Search, Cycle, Agents) ->
  receive
    {get_view, R, ReqCycle} ->
      if
        ReqCycle == Cycle ->
	  R ! Agents,
	  view(Search, Cycle, Agents);
	true ->
	  NewAgents = build_view(?OFFSETS, Search, []),
	  R ! NewAgents,
	  view(Search, ReqCycle, NewAgents)
      end;
    shutdown ->
      nop
  end.

location_loop(Loc, Neighbours, View, Agents) ->
  receive
    {enter, R, Info} ->
      {ID, _, Pos} = Info,
      Local = in_bounds(Pos),
      if
        Local -> 
	  NewInfo = {ID, Loc, Pos},
	  R ! {stay_here, NewInfo, View},
	  location_loop(Loc, Neighbours, View, [NewInfo|Agents]);
	true ->
	  {Dir, V} = calculate_direction(Pos),
	  Neighbour = get_neighbour(Dir, Neighbours),
	  NewInfo = {ID, Loc, vector_sub(Pos, make_nv(V))},
	  R ! {go_there, Neighbour, NewInfo},
	  location_loop(Loc, Neighbours, View, Agents)
      end;
    {move, R, ID, Vector} ->
      {value, Info, NewAgents} = keytake(ID, 1, Agents),
      {_, _, Pos} = Info,
      NewPos = vector_add(Pos, Vector),
      Local = in_bounds(NewPos),
      if
        Local ->
	  NewInfo = {ID, Loc, NewPos},
	  R ! {stay_here, NewInfo, View},
	  location_loop(Loc, Neighbours, View, [NewInfo|NewAgents]);
	true ->
	  {Dir, V} = calculate_direction(NewPos),
	  Neighbour = get_neighbour(Dir, Neighbours),
	  NewInfo = {ID, Loc, vector_sub(NewPos, make_nv(V))},
	  R ! {go_there, Neighbour, NewInfo},
	  location_loop(Loc, Neighbours, View, NewAgents)
      end;
    {get_info, R} ->
      R ! {info, Agents},
      location_loop(Loc, Neighbours, View, Agents);
    shutdown ->
      View ! shutdown
  end.

location(Loc) ->
  receive
    {neighbours, Neighbours} ->
      Search = reverse([self()|reverse(tuple_to_list(Neighbours))]),
      View = spawn(agents, view, [Search, -1, []]),
      location_loop(Loc, Neighbours, View, [])
  end.


%% Agents

a_sqrt(_, R, 16) -> 
  R;
a_sqrt(X, R, S) ->
  NewR = (R + (X div R)) bsr 1,
  if
    NewR == R -> R;
    true      -> a_sqrt(X, NewR, S + 1)
  end.

min(A, B) ->
  if
    A < B -> A;
    true  -> B
  end.

calculate_r(DX2, DY2, R2, DX, DY) ->
  if
    DX2 > DY2 -> a_sqrt(R2, abs(DX), 0);
    DY2 > 0   -> a_sqrt(R2, abs(DY), 0);
    true      -> 1
  end.

force_vector(Cycle, DX2, DY2, R2, DX, DY, M) ->
  if
    R2 == 0 ->
      X = Cycle band 1,
      Y = (Cycle bsr 1) band 1,
      { M * (X * (-1)), M * (Y * (-1)) };
    true    ->
      R = calculate_r(DX2, DY2, R2, DX, DY),
      { (M * DX) div R, (M * DY) div R }
  end.

calculate_force(_, _, _, _, _, [], F) ->
  F;
calculate_force(Cycle, ID, PX, PY, Pos, [A|AS], F) ->
  {AgentID, _, AgentPos} = A,
  if
    AgentID == ID ->
      calculate_force(Cycle, ID, PX, PY, Pos, AS, F);
    true          ->
      {DoX, DoY}  = vector_sub(Pos, AgentPos),
      DX          = (DoX * (PX + 1)) div 128,
      DY          = (DoY * (PY + 1)) div 128,
      DX2         = DX * DX,
      DY2         = DY * DY,
      R2          = DX2 + DY2,
      M           = min(?LOC_SIZE, ((3 * ?LOC_AREA) div (R2 + 1))),
      AddF        = force_vector(Cycle, DX2, DY2, R2, DX, DY, M),
      calculate_force(Cycle, ID, PX, PY, Pos, AS, vector_add(F, AddF))
  end.

calculate_move(Cycle, {ID, _, Pos}, Persona, View) ->
  View ! {get_view, self(), Cycle},
  receive
    Agents ->
      NewPersona = Persona + length(Agents),
      PX         = NewPersona band 255,
      PY         = (NewPersona bsr 8) band 255,
      F          = calculate_force(Cycle, ID, PX, PY, Pos, Agents, {0, 0}),
      {NewPersona, F}
  end.

update_persona({_, Loc, {X, Y}}, Persona) ->
  (((((Persona band 65535) * X) band 65535) * Y) band 65535) * Loc.

enter_agent(Info, Location) ->
  Location ! {enter, self(), Info},
  receive
    {stay_here, NewInfo, View} ->
      {NewInfo, Location, View};
    {go_there, NewLocation, NewInfo} ->
      enter_agent(NewInfo, NewLocation)
  end.

move_agent({ID, _, _}, Vector, Location) ->
  Location ! {move, self(), ID, Vector},
  receive
    {stay_here, NewInfo, View} ->
      {NewInfo, Location, View};
    {go_there, NewLocation, NewInfo} ->
      enter_agent(NewInfo, NewLocation)
  end.

agent_start(Cycle, {ID, _, _}) ->
  io:fwrite("~b ~b start~n", [Cycle, ID]).
agent_at(Cycle, {ID, Loc, {X, Y}}) ->
  io:fwrite("~b ~b at ~b:~b,~b~n", [Cycle, ID, Loc, X, Y]).
agent_end(Cycle, {ID, _, _}, Persona) ->
  io:fwrite("~b ~b end ~b~n", [Cycle, ID, Persona]).

agent_loop(Cycle, Info, Persona, Location, View) ->
  if
    Cycle < ?N_CYCLES ->
      %agent_at(Cycle, Info),
      NewCycle = Cycle + 1,
      {NewPersona, Move} = calculate_move(NewCycle, Info, update_persona(Info, Persona), View),
      sync_barrier(),
      {NewInfo, NewLocation, NewView} = move_agent(Info, Move, Location),
      sync_barrier(),
      agent_loop(NewCycle, NewInfo, NewPersona, NewLocation, NewView);
    true ->
      agent_at(Cycle, Info),
      agent_end(Cycle, Info, Persona),
      sync_barrier()
  end.

agent(Cycle, Info, Location) ->
  {ID, _, _} = Info,
  NewPersona = ID * 37,
  agent_start(Cycle, Info),
  sync_barrier(),
  {NewInfo, NewLocation, NewView} = enter_agent(Info, Location),
  sync_barrier(),
  agent_loop(Cycle, NewInfo, NewPersona, NewLocation, NewView).


%% The World

wrap(N,L) ->
  if
    N < 0 -> wrap(N + L, L);
    N < L -> N;
    true  -> N rem L
  end.

spawn_world(0, W) ->
  W;
spawn_world(S, W) ->
  Location = spawn(agents, location, [S - 1]),
  spawn_world(S - 1, [Location|W]).

build_neighbours(_, _, _, [], Neighbours) ->
  list_to_tuple(reverse(Neighbours));
build_neighbours(World_Size, World, Loc, [O|OS], LS) ->
  {OX, OY} = O,
  X        = Loc rem World_Size,
  Y        = Loc div World_Size,
  NX       = wrap(X + OX, World_Size),
  NY       = wrap(Y + OY, World_Size),
  N        = (NY * World_Size) + NX,
  Location = nth(N + 1, World),
  build_neighbours(World_Size, World, Loc, OS, [Location|LS]).

add_agents(_, _, ID, 0) ->
  ID;
add_agents(Loc_Agents, Location, ID, N) ->
  P = Loc_Agents - N,
  O = ?LOC_MIN + ((?LOC_SIZE div (Loc_Agents + 4)) * (P + 2)),
  Info = {ID, -1, {O, O}},
  spawn(agents, agent, [0, Info, Location]),
  add_agents(Loc_Agents, Location, ID + 1, N - 1).

setup_world(_, _, _, _, BaseID, []) ->
  BaseID;
setup_world(WS, LA, World, Loc, BaseID, [L|LS]) ->
  [_|Offsets] = reverse(?OFFSETS),
  L ! {neighbours, build_neighbours(WS, World, Loc, reverse(Offsets), [])},
  NextID = add_agents(LA, L, BaseID, LA),
  setup_world(WS, LA, World, Loc + 1, NextID, LS).

teardown_world([]) ->
  nop;
teardown_world([L|LS]) ->
  L ! shutdown,
  teardown_world(LS).

ticker(0) ->
  sync_barrier (),
  nop;
ticker(Count) ->
  %io:fwrite("~b A~n", [Count]),
  sync_barrier (),
  %io:fwrite("~b B~n", [Count]),
  sync_barrier (),
  ticker(Count - 1).

agents(World_Size, Loc_Agents) ->
  World_Area = World_Size * World_Size,
  init_barrier((Loc_Agents * World_Area) + 1),
  World = spawn_world(World_Area, []),
  setup_world(World_Size, Loc_Agents, World, 0, 1, World),
  ticker(?N_CYCLES + 1),
  teardown_world(World),
  shutdown_barrier(),
  erlang:halt().

main() ->
  agents(?WORLD_SIZE, ?LOC_AGENTS).
main([A1|[A2|_]]) ->
  World_Size = list_to_integer(A1),
  Loc_Agents = list_to_integer(A2),
  agents(World_Size, Loc_Agents).

File: trunk/tests/ccsp-comparisons/go/agents.go
package main

import (
	"fmt";
	"flag";
	"os";
	"runtime";
	"strconv";
)

const (
	N_CYCLES	= 1024;
	WORLD_SIZE	= 5;

	LOC_SIZE	= 4096;
	LOC_AREA	= (LOC_SIZE * LOC_SIZE);
	LOC_AGENTS	= 12;

	LOC_MAX	= ((LOC_SIZE / 2) - 1);
	LOC_MIN	= (-(LOC_SIZE / 2));
)

type Vector struct {
	x, y int;
}

type AgentInfo struct {
	id	int;
	loc	int;
	pos	Vector;
}

type AgentList []*AgentInfo
type AgentMap map[int]*AgentInfo

const (
	LOC_ENTER	= iota;
	LOC_MOVE;
	LOC_GET_VIEW;
	LOC_BORROW_INFO;
	LOC_SHUTDOWN;
	LOC_STAY_HERE;
	LOC_GO_THERE;
)

type LocationResp struct {
	cmd	int;
	loc	chan *LocationReq;
	view	chan *ViewReq;
	agents	*AgentMap;
}

type LocationReq struct {
	cmd	int;
	v	*Vector;
	info	*AgentInfo;
	resp	chan *LocationResp;
}

type ViewReq struct {
	cycle	int;
	resp	chan *AgentList;
}

const NEIGHBOURS = 8

var offsets [9]Vector = [9]Vector{
	Vector{-1, -1},
	Vector{ 0, -1},
	Vector{ 1, -1},
	Vector{-1,  0},
	Vector{ 1,  0},
	Vector{-1,  1},
	Vector{ 0,  1},
	Vector{ 1,  1},
	Vector{ 0,  0},
}

func (v *Vector) vector_add(a *Vector) {
	v.x = v.x + a.x;
	v.y = v.y + a.y;
}

func (v *Vector) vector_sub(a *Vector) {
	v.x = v.x - a.x;
	v.y = v.y - a.y;
}

func dir_value(n int) int {
	if n < LOC_MIN {
		return -1
	} else if n > LOC_MAX {
		return 1
	}
	return 0;
}

func (p Vector) in_bounds() bool {
	if p.x < LOC_MIN || p.x > LOC_MAX {
		return true
	}
	return !(p.y < LOC_MIN || p.y > LOC_MAX);
}

func a_sqrt(x, r int) int {
	last_r := 0;
	steps := 0;
	for (r != last_r) && (steps < 16) {
		last_r, r, steps = r, (r+(x/r))>>1, steps+1
	}
	return r;
}

type Barrier chan BarrierReq
type BarrierReq chan int
type BarrierEnd struct {
	barrier	Barrier;
	resp	BarrierReq;
}

func barrier_runner(b Barrier, count int) {
	waiting_list := make([]BarrierReq, count);
	for {
		for n := 0; n < count; n = n + 1 {
			rc := <-b;
			if rc != nil {
				waiting_list[n] = rc
			} else {
				return
			}
		}
		for n := 0; n < count; n = n + 1 {
			waiting_list[n] <- n
		}
	}
}

func new_barrier(count int) Barrier {
	b := make(Barrier, count / 16);
	go barrier_runner(b, count);
	return b;
}

func new_barrier_end(b Barrier) BarrierEnd {
	return BarrierEnd{b, make(BarrierReq)}
}

func (b *BarrierEnd) sync() {
	b.barrier <- b.resp;
	<-b.resp;
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x;
}

func agent_start(log chan string, cycle int, info *AgentInfo) {
	log <- fmt.Sprintf("%d %d start\n", cycle, info.id)
}
func agent_at(log chan string, cycle int, info *AgentInfo) {
	log <- fmt.Sprintf("%d %d at %d:%d,%d\n", cycle, info.id, info.loc, info.pos.x, info.pos.y)
}
func agent_end(log chan string, cycle, persona int, info *AgentInfo) {
	log <- fmt.Sprintf("%d %d end %d\n", cycle, info.id, persona)
}

func agent(cycle int, info AgentInfo, loc chan *LocationReq, bar BarrierEnd, log chan string) {
	persona := info.id * 37;
	loc_resp := make(chan *LocationResp);
	view_resp := make(chan *AgentList);

	agent_start(log, cycle, &info);

	bar.sync();

	loc <- &LocationReq{LOC_ENTER, nil, &info, loc_resp};
	msg := <-loc_resp;
	view_req := msg.view;

	bar.sync();

	for cycle < N_CYCLES {
		persona = (persona & 65535) * info.pos.x;
		persona = (persona & 65535) * info.pos.y;
		persona = (persona & 65535) * info.loc;

		// agent_at (log, cycle, &info);

		cycle = cycle + 1;

		view_req <- &ViewReq{cycle, view_resp};
		agents := <-view_resp;

		persona = persona + len(*agents);
		px	:= persona & 0xff;
		py	:= (persona >> 8) & 0xff;
		force	:= Vector{0, 0};

		for _, oa := range *agents {
			if oa.id == info.id {
				continue
			}

			d := info.pos;
			d.vector_sub(&oa.pos);
			d.x = (d.x * (px + 1)) / 128;
			d.y = (d.y * (py + 1)) / 128;
			dx2 := d.x * d.x;
			dy2 := d.y * d.y;
			r2 := dx2 + dy2;
			f := (3*(LOC_SIZE*LOC_SIZE)) / (r2+1);
			if f > LOC_SIZE {
				f = LOC_SIZE;
			}

			r := 1;
			if dx2 > dy2 {
				r = a_sqrt(r2, abs(d.x))
			} else if dy2 > 0 {
				r = a_sqrt(r2, abs(d.y))
			}

			if r2 != 0 {
				force.x = force.x + ((f * d.x) / r);
				force.y = force.y + ((f * d.y) / r);
			} else {
				x_jitter := (cycle & 1) * (-1);
				y_jitter := ((cycle >> 1) & 1) * (-1);
				force.x = force.x + (f * x_jitter);
				force.y = force.y + (f * y_jitter);
			}
		}

		bar.sync();

		loc <- &LocationReq{LOC_MOVE, &force, &info, loc_resp};
		msg := <-loc_resp;

		switch msg.cmd {
		case LOC_STAY_HERE:
			/* nop */
		case LOC_GO_THERE:
			loc = msg.loc;
			for moving := true; moving; {
				loc <- &LocationReq{LOC_ENTER, nil, &info, loc_resp};
				msg := <-loc_resp;

				switch msg.cmd {
				case LOC_STAY_HERE:
					view_req = msg.view;
					moving = false;
				case LOC_GO_THERE:
					loc = msg.loc
				}
			}
		}

		bar.sync();
	}

	bar.sync();

	agent_at(log, cycle, &info);
	agent_end(log, cycle, persona, &info);

	bar.sync();
}

func viewer(req chan *ViewReq, fov []chan *LocationReq) {
	cycle := -1;
	view := make(AgentList, 0, 64);
	loc_resp := make(chan *LocationResp);

	done := false;
	for !done {
		msg := <-req;
		if msg.cycle >= 0 {
			if msg.cycle != cycle {
				view = view[0:0];
				for dir, loc := range (fov) {
					offset := Vector{LOC_SIZE * offsets[dir].x, LOC_SIZE * offsets[dir].y};
					loc <- &LocationReq{LOC_BORROW_INFO, nil, nil, loc_resp};
					msg := <-loc_resp;
					if len(view)+len(*msg.agents) > cap(view) {
						new_view := make(AgentList, len(view), (len(view)+len(*msg.agents))*2);
						for i, a := range view {
							new_view[i] = a
						}
						view = new_view;
					}
					view_len := len(view);
					view = view[0:(view_len + len(*msg.agents))];
					for _, a := range *msg.agents {
						agent := *a;
						agent.pos.vector_add(&offset);
						view[view_len] = &agent;
						view_len = view_len + 1;
					}
					loc_resp <- msg;
				}
				cycle = msg.cycle;
			}
			msg.resp <- &view;
		} else {
			done = true
		}
	}
}

func (info *AgentInfo) handle_update() (bool, int) {
	dx := dir_value(info.pos.x);
	dy := dir_value(info.pos.y);
	if (dx != 0) || (dy != 0) {
		for d, v := range offsets {
			if (dx == v.x) && (dy == v.y) {
				info.pos.x = info.pos.x - (v.x * LOC_SIZE);
				info.pos.y = info.pos.y - (v.y * LOC_SIZE);
				return true, d;
			}
		}
	}
	return false, 0;
}

func location(loc int, req chan *LocationReq, neighbours []chan *LocationReq) {
	view := make(chan *ViewReq, 16);
	go viewer(view, neighbours);

	agents := make(AgentMap, 16);

	done := false;
	for !done {
		msg := <-req;
		switch msg.cmd {
		case LOC_ENTER:
			moved, dir := msg.info.handle_update();
			if moved {
				msg.resp <- &LocationResp{LOC_GO_THERE, neighbours[dir], nil, nil}
			} else {
				info := msg.info;
				info.loc = loc;
				agents[info.id] = info;
				msg.resp <- &LocationResp{LOC_STAY_HERE, nil, view, nil};
			}
		case LOC_MOVE:
			info := msg.info;
			info.pos.vector_add(msg.v);
			moved, dir := info.handle_update();
			if moved {
				agents[info.id] = nil, false;
				msg.resp <- &LocationResp{LOC_GO_THERE, neighbours[dir], nil, nil};
			} else {
				msg.resp <- &LocationResp{LOC_STAY_HERE, nil, view, nil}
			}
		case LOC_BORROW_INFO:
			msg.resp <- &LocationResp{0, nil, nil, &agents};
			<-msg.resp;
		case LOC_SHUTDOWN:
			done = true
		}
	}

	view <- &ViewReq{-1, nil};
}

func wrap(n, max int) int {
	for n < 0 {
		n = n + max
	}
	return n % max;
}

func agents(world_size, loc_agents int, screen chan string) {
	world_area	:= world_size * world_size;
	world		:= make([]chan *LocationReq, world_area);

	for loc := 0; loc < world_area; loc = loc + 1 {
		world[loc] = make(chan *LocationReq, 16)
	}
	for loc := 0; loc < world_area; loc = loc + 1 {
		this_loc	:= world[loc];
		x		:= loc % world_size;
		y		:= loc / world_size;
		neighbours	:= make([]chan *LocationReq, NEIGHBOURS + 1);
		for i := 0; i < NEIGHBOURS; i = i + 1 {
			n_x := wrap(x+offsets[i].x, world_size);
			n_y := wrap(y+offsets[i].y, world_size);
			neighbours[i] = world[(n_y*world_size)+n_x];
		}
		neighbours[NEIGHBOURS] = this_loc;
		go location(loc, this_loc, neighbours);
	}

	b := new_barrier((loc_agents * world_area) + 1);

	id := 1;
	for loc := 0; loc < world_area; loc = loc + 1 {
		for p := 0; p < loc_agents; p = p + 1 {
			x := LOC_MIN + ((LOC_SIZE / (loc_agents + 4)) * (p + 2));
			y := LOC_MIN + ((LOC_SIZE / (loc_agents + 4)) * (p + 2));
			info := AgentInfo{id, loc, Vector{x, y}};
			go agent(0, info, world[loc], new_barrier_end(b), screen);
			id = id + 1;
		}
	}

	be := new_barrier_end(b);
	for cycle := 0; cycle <= N_CYCLES; cycle = cycle + 1 {
		be.sync();
		be.sync();
	}
	be.sync();
	be.sync();
	b <- nil;

	for loc := 0; loc < world_area; loc = loc + 1 {
		world[loc] <- &LocationReq{LOC_SHUTDOWN, nil, nil, nil}
	}
}

func screen_writer(in chan string) {
	for {
		str := <-in;
		if len(str) == 0 {
			return
		}
		os.Stdout.WriteString(str)
	}
}

func main() {
	flag.Parse();

	threads, _ := strconv.Atoi(os.Getenv("CORES"));
	if threads < 1 {
		threads = 1
	}
	runtime.GOMAXPROCS(threads);

	world_size := WORLD_SIZE;
	if flag.NArg() >= 1 {
		world_size, _ = strconv.Atoi(flag.Arg(0))
	}
	loc_agents := LOC_AGENTS;
	if flag.NArg() >= 2 {
		loc_agents, _ = strconv.Atoi(flag.Arg(1))
	}

	screen := make(chan string, 16);
	go screen_writer(screen);
	agents(world_size, loc_agents, screen);
	screen <- ""
}
